src/stream/mod.rs

```diff
@@ -2,3 +2,49 @@
 
 pub mod log;
 pub mod stdin;
+
+use crate::{Captures, Needle, Error};
+use std::io;
+
+#[cfg(not(feature = "async"))]
+use std::io::{Write, Read, BufRead};
+
+#[cfg(feature = "async")]
+use futures_lite::{AsyncRead, AsyncWrite, AsyncBufRead};
+
+/// Trait for types that can read and write to child programs.
+#[cfg(not(feature = "async"))]
+pub trait StreamSink: Write + Read + BufRead {
```

Comment:
I was aproaching something like this trait several times as I remember and got not a lot of success :sweat_smile: 

I was thinking that wouldn't it be cool to have a `Session` as a trait.
So we would not need all these generics in the parameters just `impl Session`/`impl Expect`.

But as I remember something was wrong with a rust typesystem checks in async.

So I guess it's valuable trait.
But it may require some design changes/thinking.

Reply:
I think this trait works and we should implement this trait for other types such as `Session` (which I haven't currently done). The difficult part is the naming, maybe simply `Expect` rather than `StreamSink`?

Reply:
> I think this trait works and we should implement this trait for other types such as Session

Exactly;

But it's not a easy to do change; needs some thinking/refactorings

Reply:
But why do we need separate traits? As far as I understand the code anywhere we can expect we can also send, in effect the expect and send functions are really just higher level abstractions for `ReadBuf`, `Write` etc. or am I missing a particular reason they should be distinct?